In [ ]:
#Using python 2.7, install tensorflow/networkx
!pip install tensorflow==1.9
!pip install networkx==1.11

     |████████████████████████████████| 51.2MB 67kB/s 
     |████████████████████████████████| 3.3MB 51.8MB/s 
     |████████████████████████████████| 573kB 48.7MB/s 
ERROR: google-auth 1.27.0 has requirement setuptools>=40.3.0, but you'll have setuptools 39.1.0 which is incompatible.
  Found existing installation: tensorboard 2.1.0
    Uninstalling tensorboard-2.1.0:
      Successfully uninstalled tensorboard-2.1.0
  Found existing installation: setuptools 44.1.1
    Uninstalling setuptools-44.1.1:
      Successfully uninstalled setuptools-44.1.1
  Found existing installation: tensorflow 1.3.0
    Uninstalling tensorflow-1.3.0:
      Successfully uninstalled tensorflow-1.3.0


## Testing Code
Here we can test code by running the bash commands, like:
python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.5



In [ ]:

!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.5


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.81494 train_acc= 0.69646 val_roc= 0.64483 val_ap= 0.66303 test_roc= 0.67086 test_ap= 0.69082
Epoch: 0011 train_loss= 0.77914 train_acc= 0.73076 val_roc= 0.69069 val_ap= 0.72648 test_roc= 0.74284 test_ap= 0.77705
Epoch: 0021 train_loss= 0.72185 train_acc= 0.73859 val_roc= 0.69867 val_ap= 0.73546 test_roc= 0.74208 test_ap= 0.77315
Epoch: 0031 train_loss= 0.66953 train_acc= 0.79938 val_roc= 0.68753 val_ap= 0.71592 test_roc= 0.71839 test_ap= 0.74579
Epoch: 0041 train_loss= 0.63636 train_acc= 0.72352 val_roc= 0.66645 val_ap= 0.70332 test_roc= 0.67162 test_ap= 0.71338
Epoch: 0051 train_loss= 0.60754 train_acc= 0.74111 val_roc= 0.72275 val_ap= 0.74459 test_roc= 0.72817 test_ap= 0.76291
Epoch: 0061 train_loss= 0.57443 train_acc= 0.71415 val_roc= 0.79153 val_ap= 0.80903 test_roc= 0.76920 test_ap= 0.80380
Epoch: 0071 train_loss= 0.53979 train_acc= 0.67373 val_roc= 0.82536 val_ap= 0.84447 test_roc= 0.80133 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap, lambda=0
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 1.70175 train_acc= 0.55239 val_roc= 0.70905 val_ap= 0.70852 test_roc= 0.70745 test_ap= 0.71996
Epoch: 0011 train_loss= 1.30582 train_acc= 0.56647 val_roc= 0.72009 val_ap= 0.74398 test_roc= 0.75390 test_ap= 0.77427
Epoch: 0021 train_loss= 0.79658 train_acc= 0.60153 val_roc= 0.71681 val_ap= 0.74445 test_roc= 0.74812 test_ap= 0.77533
Epoch: 0031 train_loss= 0.70021 train_acc= 0.68991 val_roc= 0.71661 val_ap= 0.74097 test_roc= 0.74154 test_ap= 0.77232
Epoch: 0041 train_loss= 0.65052 train_acc= 0.77198 val_roc= 0.71753 val_ap= 0.73497 test_roc= 0.72746 test_ap= 0.75784
Epoch: 0051 train_loss= 0.60729 train_acc= 0.69033 val_roc= 0.72787 val_ap= 0.73857 test_roc= 0.72597 test_ap= 0.75450
Epoch: 0061 train_loss= 0.53662 train_acc= 0.68595 val_roc= 0.81505 val_ap= 0.83722 test_roc= 0.80901 test_ap= 0.83941
Epoch: 0071 train_loss= 0.50974 train_acc= 0.65647 val_roc= 0.85011 val_ap= 0.86694 test_roc= 0.83843 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap, lambda=0.1
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.1


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 1.42890 train_acc= 0.56506 val_roc= 0.68839 val_ap= 0.65700 test_roc= 0.70719 test_ap= 0.71382
Epoch: 0011 train_loss= 1.18472 train_acc= 0.58143 val_roc= 0.74912 val_ap= 0.75478 test_roc= 0.75193 test_ap= 0.77755
Epoch: 0021 train_loss= 0.81230 train_acc= 0.63654 val_roc= 0.73078 val_ap= 0.74344 test_roc= 0.75556 test_ap= 0.77611
Epoch: 0031 train_loss= 0.70377 train_acc= 0.72533 val_roc= 0.72720 val_ap= 0.73515 test_roc= 0.74787 test_ap= 0.76641
Epoch: 0041 train_loss= 0.64148 train_acc= 0.79398 val_roc= 0.68703 val_ap= 0.71749 test_roc= 0.69166 test_ap= 0.73114
Epoch: 0051 train_loss= 0.59609 train_acc= 0.70733 val_roc= 0.72373 val_ap= 0.74447 test_roc= 0.71542 test_ap= 0.74853
Epoch: 0061 train_loss= 0.54037 train_acc= 0.68311 val_roc= 0.81234 val_ap= 0.82910 test_roc= 0.79816 test_ap= 0.80850
Epoch: 0071 train_loss= 0.51103 train_acc= 0.66812 val_roc= 0.83783 val_ap= 0.85726 test_roc= 0.83372 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap, lambda = 0.2
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 1.72197 train_acc= 0.55286 val_roc= 0.67861 val_ap= 0.69482 test_roc= 0.68627 test_ap= 0.69730
Epoch: 0011 train_loss= 1.40611 train_acc= 0.56254 val_roc= 0.71160 val_ap= 0.73191 test_roc= 0.76564 test_ap= 0.77860
Epoch: 0021 train_loss= 0.88782 train_acc= 0.57697 val_roc= 0.71642 val_ap= 0.73557 test_roc= 0.76437 test_ap= 0.77757
Epoch: 0031 train_loss= 0.71443 train_acc= 0.67583 val_roc= 0.71920 val_ap= 0.74025 test_roc= 0.75617 test_ap= 0.77010
Epoch: 0041 train_loss= 0.67102 train_acc= 0.75798 val_roc= 0.71798 val_ap= 0.73508 test_roc= 0.73777 test_ap= 0.75366
Epoch: 0051 train_loss= 0.61536 train_acc= 0.70884 val_roc= 0.72266 val_ap= 0.74015 test_roc= 0.72636 test_ap= 0.75201
Epoch: 0061 train_loss= 0.55995 train_acc= 0.67804 val_roc= 0.78052 val_ap= 0.79915 test_roc= 0.78572 test_ap= 0.81734
Epoch: 0071 train_loss= 0.53197 train_acc= 0.66191 val_roc= 0.82161 val_ap= 0.82891 test_roc= 0.81295 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap, lambda = 0.3
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.3


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 1.04724 train_acc= 0.60481 val_roc= 0.64034 val_ap= 0.64349 test_roc= 0.68936 test_ap= 0.67930
Epoch: 0011 train_loss= 0.92539 train_acc= 0.62628 val_roc= 0.72319 val_ap= 0.75619 test_roc= 0.75032 test_ap= 0.77945
Epoch: 0021 train_loss= 0.75196 train_acc= 0.64142 val_roc= 0.72222 val_ap= 0.74972 test_roc= 0.75182 test_ap= 0.77712
Epoch: 0031 train_loss= 0.67614 train_acc= 0.74811 val_roc= 0.71010 val_ap= 0.73360 test_roc= 0.73415 test_ap= 0.75550
Epoch: 0041 train_loss= 0.63736 train_acc= 0.75469 val_roc= 0.68249 val_ap= 0.71616 test_roc= 0.68248 test_ap= 0.72021
Epoch: 0051 train_loss= 0.60506 train_acc= 0.72579 val_roc= 0.73456 val_ap= 0.75277 test_roc= 0.72948 test_ap= 0.75673
Epoch: 0061 train_loss= 0.55528 train_acc= 0.70824 val_roc= 0.80260 val_ap= 0.82507 test_roc= 0.79095 test_ap= 0.81770
Epoch: 0071 train_loss= 0.51954 train_acc= 0.66875 val_roc= 0.82867 val_ap= 0.85375 test_roc= 0.82967 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap, lambda = 0.4
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.4


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.90259 train_acc= 0.64199 val_roc= 0.66903 val_ap= 0.67100 test_roc= 0.68656 test_ap= 0.69007
Epoch: 0011 train_loss= 0.84801 train_acc= 0.66758 val_roc= 0.76150 val_ap= 0.77715 test_roc= 0.75238 test_ap= 0.77144
Epoch: 0021 train_loss= 0.73597 train_acc= 0.71688 val_roc= 0.72206 val_ap= 0.75274 test_roc= 0.74176 test_ap= 0.76784
Epoch: 0031 train_loss= 0.68723 train_acc= 0.79369 val_roc= 0.71784 val_ap= 0.73967 test_roc= 0.73420 test_ap= 0.75464
Epoch: 0041 train_loss= 0.63443 train_acc= 0.78718 val_roc= 0.70050 val_ap= 0.72766 test_roc= 0.69746 test_ap= 0.73251
Epoch: 0051 train_loss= 0.59193 train_acc= 0.71634 val_roc= 0.75103 val_ap= 0.77250 test_roc= 0.73864 test_ap= 0.77175
Epoch: 0061 train_loss= 0.55021 train_acc= 0.67051 val_roc= 0.81697 val_ap= 0.82637 test_roc= 0.80084 test_ap= 0.82237
Epoch: 0071 train_loss= 0.52175 train_acc= 0.68055 val_roc= 0.84324 val_ap= 0.85513 test_roc= 0.84329 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap lambda=0.5
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.5


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.80882 train_acc= 0.69771 val_roc= 0.67574 val_ap= 0.66588 test_roc= 0.67361 test_ap= 0.67749
Epoch: 0011 train_loss= 0.78130 train_acc= 0.72602 val_roc= 0.74086 val_ap= 0.76179 test_roc= 0.77725 test_ap= 0.79557
Epoch: 0021 train_loss= 0.73778 train_acc= 0.77119 val_roc= 0.70055 val_ap= 0.73771 test_roc= 0.74317 test_ap= 0.77417
Epoch: 0031 train_loss= 0.69061 train_acc= 0.74512 val_roc= 0.70733 val_ap= 0.73260 test_roc= 0.73703 test_ap= 0.76095
Epoch: 0041 train_loss= 0.63045 train_acc= 0.76405 val_roc= 0.70089 val_ap= 0.73006 test_roc= 0.68503 test_ap= 0.72430
Epoch: 0051 train_loss= 0.60062 train_acc= 0.69555 val_roc= 0.74014 val_ap= 0.76289 test_roc= 0.71007 test_ap= 0.74698
Epoch: 0061 train_loss= 0.56031 train_acc= 0.69319 val_roc= 0.79488 val_ap= 0.81152 test_roc= 0.77057 test_ap= 0.79144
Epoch: 0071 train_loss= 0.54066 train_acc= 0.66329 val_roc= 0.81677 val_ap= 0.82781 test_roc= 0.79141 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap lambda=0.6
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.6


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.76083 train_acc= 0.79032 val_roc= 0.61789 val_ap= 0.62881 test_roc= 0.63921 test_ap= 0.65157
Epoch: 0011 train_loss= 0.74918 train_acc= 0.81098 val_roc= 0.71679 val_ap= 0.74306 test_roc= 0.75284 test_ap= 0.78539
Epoch: 0021 train_loss= 0.73716 train_acc= 0.84528 val_roc= 0.69972 val_ap= 0.74628 test_roc= 0.73555 test_ap= 0.77489
Epoch: 0031 train_loss= 0.70289 train_acc= 0.79552 val_roc= 0.70044 val_ap= 0.73546 test_roc= 0.72951 test_ap= 0.76502
Epoch: 0041 train_loss= 0.65214 train_acc= 0.77438 val_roc= 0.70247 val_ap= 0.72256 test_roc= 0.70108 test_ap= 0.73918
Epoch: 0051 train_loss= 0.61787 train_acc= 0.73457 val_roc= 0.68276 val_ap= 0.71691 test_roc= 0.69032 test_ap= 0.73562
Epoch: 0061 train_loss= 0.58032 train_acc= 0.71118 val_roc= 0.73422 val_ap= 0.75769 test_roc= 0.75306 test_ap= 0.79217
Epoch: 0071 train_loss= 0.54352 train_acc= 0.68512 val_roc= 0.76657 val_ap= 0.78204 test_roc= 0.80084 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap lambda=0.7
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.7


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.74433 train_acc= 0.91735 val_roc= 0.63390 val_ap= 0.65983 test_roc= 0.62630 test_ap= 0.64848
Epoch: 0011 train_loss= 0.74360 train_acc= 0.88814 val_roc= 0.67291 val_ap= 0.71574 test_roc= 0.69969 test_ap= 0.73603
Epoch: 0021 train_loss= 0.74001 train_acc= 0.85556 val_roc= 0.67423 val_ap= 0.71484 test_roc= 0.69924 test_ap= 0.73492
Epoch: 0031 train_loss= 0.72028 train_acc= 0.84558 val_roc= 0.68052 val_ap= 0.71214 test_roc= 0.71236 test_ap= 0.73734
Epoch: 0041 train_loss= 0.66888 train_acc= 0.79960 val_roc= 0.68775 val_ap= 0.71513 test_roc= 0.70237 test_ap= 0.72873
Epoch: 0051 train_loss= 0.63700 train_acc= 0.71768 val_roc= 0.69202 val_ap= 0.72161 test_roc= 0.68409 test_ap= 0.71991
Epoch: 0061 train_loss= 0.61405 train_acc= 0.68762 val_roc= 0.72816 val_ap= 0.73915 test_roc= 0.73820 test_ap= 0.75818
Epoch: 0071 train_loss= 0.60483 train_acc= 0.67016 val_roc= 0.74384 val_ap= 0.75496 test_roc= 0.75840 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap lambda=0.8
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.8


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.76185 train_acc= 0.99648 val_roc= 0.60083 val_ap= 0.61264 test_roc= 0.63431 test_ap= 0.64188
Epoch: 0011 train_loss= 0.75249 train_acc= 0.94813 val_roc= 0.67210 val_ap= 0.70430 test_roc= 0.71059 test_ap= 0.75116
Epoch: 0021 train_loss= 0.74481 train_acc= 0.89634 val_roc= 0.68366 val_ap= 0.71398 test_roc= 0.71620 test_ap= 0.74947
Epoch: 0031 train_loss= 0.71990 train_acc= 0.83926 val_roc= 0.68905 val_ap= 0.71501 test_roc= 0.71978 test_ap= 0.74894
Epoch: 0041 train_loss= 0.66929 train_acc= 0.75896 val_roc= 0.69878 val_ap= 0.72404 test_roc= 0.70205 test_ap= 0.74122
Epoch: 0051 train_loss= 0.63799 train_acc= 0.71817 val_roc= 0.67725 val_ap= 0.68820 test_roc= 0.66929 test_ap= 0.68420
Epoch: 0061 train_loss= 0.62172 train_acc= 0.69763 val_roc= 0.71101 val_ap= 0.72352 test_roc= 0.70324 test_ap= 0.72127
Epoch: 0071 train_loss= 0.61576 train_acc= 0.67571 val_roc= 0.70764 val_ap= 0.72282 test_roc= 0.71483 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap lamdba = 0.9
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 0.9


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.78699 train_acc= 0.99841 val_roc= 0.63502 val_ap= 0.66323 test_roc= 0.67740 test_ap= 0.71324
Epoch: 0011 train_loss= 0.77802 train_acc= 0.99758 val_roc= 0.68757 val_ap= 0.72600 test_roc= 0.72595 test_ap= 0.76507
Epoch: 0021 train_loss= 0.77110 train_acc= 0.98857 val_roc= 0.70644 val_ap= 0.73314 test_roc= 0.73396 test_ap= 0.76822
Epoch: 0031 train_loss= 0.76405 train_acc= 0.87783 val_roc= 0.71124 val_ap= 0.73672 test_roc= 0.73476 test_ap= 0.76635
Epoch: 0041 train_loss= 0.76527 train_acc= 0.88387 val_roc= 0.71116 val_ap= 0.73696 test_roc= 0.73463 test_ap= 0.76287
Epoch: 0051 train_loss= 0.75288 train_acc= 0.83304 val_roc= 0.70663 val_ap= 0.73021 test_roc= 0.72616 test_ap= 0.75189
Epoch: 0061 train_loss= 0.74120 train_acc= 0.91046 val_roc= 0.71467 val_ap= 0.73030 test_roc= 0.73453 test_ap= 0.75549
Epoch: 0071 train_loss= 0.71377 train_acc= 0.83970 val_roc= 0.71606 val_ap= 0.72766 test_roc= 0.71483 te

In [ ]:
#Now we want to check how autoregressive-scalar changes our roc/ap lambda = 1
!bash python train.py --epochs 500 --model feedback --edge_dropout 0.5 --learning_rate 0.01 --autoregressive_scalar 1


Instructions for updating:
dim is deprecated, use axis instead
Epoch: 0001 train_loss= 0.80229 train_acc= 0.99841 val_roc= 0.62736 val_ap= 0.64227 test_roc= 0.66156 test_ap= 0.69273
Epoch: 0011 train_loss= 0.79219 train_acc= 0.99841 val_roc= 0.64032 val_ap= 0.67239 test_roc= 0.63745 test_ap= 0.66338
Epoch: 0021 train_loss= 0.78417 train_acc= 0.99841 val_roc= 0.63090 val_ap= 0.65285 test_roc= 0.64381 test_ap= 0.65973
Epoch: 0031 train_loss= 0.78345 train_acc= 0.99841 val_roc= 0.63902 val_ap= 0.66368 test_roc= 0.65629 test_ap= 0.66999
Epoch: 0041 train_loss= 0.78395 train_acc= 0.99841 val_roc= 0.63593 val_ap= 0.66507 test_roc= 0.65827 test_ap= 0.68034
Epoch: 0051 train_loss= 0.79208 train_acc= 0.99841 val_roc= 0.63266 val_ap= 0.65908 test_roc= 0.65916 test_ap= 0.68059
Epoch: 0061 train_loss= 0.78906 train_acc= 0.99841 val_roc= 0.63614 val_ap= 0.65916 test_roc= 0.65781 test_ap= 0.67829
Epoch: 0071 train_loss= 0.79097 train_acc= 0.99841 val_roc= 0.63714 val_ap= 0.65797 test_roc= 0.66140 te

## Results from testing lambda:
        update = (1 - FLAGS.autoregressive_scalar) * z + FLAGS.autoregressive_scalar * update

So lambda = 0 means VGAE/GAE, and lambda = 1 means use iterative update only, without initial Z

| lambda | roc | ap |
| --- | --- | --- |
| 1.0 | .67 | .65 |
| 0.9 | .72 | .74 |
| 0.8 | .91 | .93 |
| 0.7 | .90 | .92 |
| 0.6 | .90 | .92 |
| 0.5 | .91 | .93 |
| 0.4 | .91 | .93 |
| 0.3 | .92 | .93 |
| 0.2 | .91 | .93 |
| 0.1 | .91 | .93 |
| 0.0 | .92 | .93 |
